### DISCUSSION TITLE
SBOM Included in Keycloak Releases

### DISCUSSION THREAD

**Discussion by idunbarh** [keycloak/keycloak#31952] [Ideas]
Hi Everyone! 
I wanted to see the community's interest in [Feature - Include sbom for KeyCloak release #9448](https://github.com/keycloak/keycloak/issues/9448).

There is a large [CISA Software Bill of Materials Community](https://www.cisa.gov/sbom) and some of us are working on a Tiger Team to help create "golden paths" for SBOM generation in Open Source projects. There is a lot of overlap with OpenSSF's SBOM-Everywhere Special Interest Group.

The [SBOM Generation Tiger Team](https://github.com/CISA-SBOM-Community/SBOM-Generation) is interested in contributing high quality SBOMs to the Keycloak releases as an exemplar.

I'd love to hear feedback on the idea.

**Comment by VinodAnandan** [comment]
Hi @idunbarh Thank you for opening this discussion and restarting the previous GitHub issues. I just want to let everyone know that there is a Quarkus team effort going on to provide native SBOM support in Quarkus and offer better transparency (including pedigree, evidence, etc.). I think Alexey Loubyansky (@aloubyansky)from the Quarkus team can provide more information.

**Reply by idunbarh** [reply]
Awesome @VinodAnandan! It would be great to hear what the Quarkus team is thinking from @aloubyansky.

Part of what we're thinking is around ensuring sboms are more "complete", so it would be good to hear what is already being thought about, and what would need to be "enrichment" of additional metadata.

**Reply by aloubyansky** [reply]
Keycloak is going a bit beyond a typical Quarkus app by adding an admin CLI "on a side" (JARs + scripts) and a few other files (text ones in the default distribution). Since this content is added after a Quarkus build, Quarkus can't include it in SBOMs it produces.
There could be a couple of ways to generate SBOMs for Keycloak distributions:

1) generate one for the server module using a (to be released) Quarkus extension; then generate one for the CLI and merge them. It may be worth also including the rest of the content found in distribution archives.
2) Enhance the Quarkus SBOM generator to cover the use-case of manifesting an already built mutable-jar distribution with "extra" content around.

Depending on what the requirements are for the final SBOM, I think both options could work.

**Reply by aloubyansky** [reply]
Given that Keycloak actually offers an admin/config capability, with the possibility of customizing the distribution, including adding custom providers, etc, they may want to consider including an option to generate an SBOM for a distribution currently configured. That would basically take the Quarkus SBOM and manifest the rest of content on top of that.

**Comment by stianst** [comment]
Doing an sbom for java dependencies only is pretty straightforward and there's decent tools to do this, but our problem is we have both Java/maven and pnpm dependencies, and have not found a nice way to do both.

I see from your project you are not planning anything for pnpm?

**Reply by idunbarh** [reply]
For keycloak to have a useful "complete" sbom, it would need to include both java and pnpm dependencies. We'd want both included. The tiger team is working on "trees of SBOMs" and dependencies between multiple languages and technologies.

Theres a couple types of problems we're looking to solve:
- delivering cyclonedx and spdx sboms that are consistent with each other
- adding top level project metadata to the sboms
- enriching with additional dependency data to met "completeness" needs
- trees of sboms to articulate dependencies across ecosystems and technology

Happy to iterate on a PR to ensure the keycloak community and maintainers are happy with the resulting sboms.

**Reply by aloubyansky** [reply]
@idunbarh could you expand a bit more on trees of SBOMs? Are there some examples of those? Thanks.

**Reply by idunbarh** [reply]
@aloubyansky Thinking more about "systems of SBOMs". 

Instead of have large SBOMs, lets decompose them into smaller SBOMs with external references between them.

For example for a helm chart, it could an SBOM with a chart and list of container images. For each component thats a container image then has a external reference to a SBOM for that specific container image.

**Reply by VinodAnandan** [reply]
Since maven and pnpm follow separate build processes, I believe it’s perfectly fine to generate SBOMs separately using different tools. For instance, cdxgen ( https://github.com/CycloneDX/cdxgen ) can handle SBOM generation for pnpm projects. If needed, these SBOMs can also be merged using the CycloneDX CLI tool (https://github.com/CycloneDX/cyclonedx-cli#merge-command). However, I think it might be more appropriate to keep them separate, as they are two distinct technologies.

**Reply by awi-MACH** [reply]
Hello all. Issue #9448 seems to be in stasis. Is anyone working on this? We're looking forward to this feature and we'd also appreciate a separation, as discussed above, if this simplifies the implementation. :-)

**Reply by aloubyansky** [reply]
There is (in the upcoming release of Quarkus) a way for Quarkus extensions to contribute components and dependencies to the Quarkus SBOM generator.
I actually have a PoC for Keycloak with Maven and npm dependencies.
It would require adding quarkus-cyclonedx as a dependency though.

The SBOM could be external, embedded and/or exposed through an endpoint (/.well-known/sbom by default).

**Reply by VinodAnandan** [reply]
@aloubyansky What is the estimated timeline for this upcoming release of Quarkus?
In my humble opinion, perfection can sometimes become the enemy of progress. I don’t think having multiple SBOM files for now or combining them externally into a single SBOM is a significant concern at this stage? It may be better to start with a practical approach and iterate from there. WDYT ?

**Reply by aloubyansky** [reply]
It's still not that straightforward in this case. We do have options. What needs to be clarified is at what degree of data accuracy this actually starts to make sense to the team and consumers.

**Comment by stianst** [comment]
Adding SBOMs are on our roadmap, and we will deliver on that. I can't say exactly when at this point though as we need to figure out what tool fits best for us (CycloneDX, cdgexn, Trivy, and even the Quarkus extension, are things we will look at). We also need to not just generate some random SBOM, but we need to verify that the actually information inside is valid at least to some extent. As an example if the generated SBOM does not include a dependency for some reason, then having an SBOM can actually be worse than scanning the distribution directly with for instance Trivy.

Couple questions though:

* For containers I assume an SBOM should also include container dependencies as well
* Should we distribute the SBOM as a separate download, or bundle it together with the distribution? That may be a bit tricky especially for the container

**Reply by aloubyansky** [reply]
In case of containers, there is an option to attach SBOMs as attestations with cosign. For example, Grype supports reading SBOMs from attestations. But not all scanners support that.
Embedding SBOMs in a distribution could generally cover more cases. Especially when customers rebuild container images and the original metadata is lost.

**Reply by aloubyansky** [reply]
I opened https://github.com/keycloak/keycloak/pull/51743 for review.
